kernel/kernel-api/src/main/java/io/delta/kernel/internal/DeltaHistoryManager.java

```diff
@@ -0,0 +1,249 @@
+            if (!files.hasNext()) {
+                // We treat an empty directory as table not found
+                throw new TableNotFoundException(logPath.getParent().toString()); /* use dataPath */
+            }
+            return files;
+        } catch (FileNotFoundException e) {
+            throw new TableNotFoundException(logPath.toString());
+        } catch (IOException io) {
+            throw new RuntimeException("Failed to list the files in delta log", io);
+        }
+    }
+
+    /**
+     * Returns the commit version and timestamps of all commits starting from version {@code start}.
+     * Guarantees that the commits returned have both monotonically increasing versions and
+     * timestamps.
+     */
+    private static List<Commit> getCommits(TableClient tableClient, Path logPath, long start)
+            throws TableNotFoundException{
+        CloseableIterator<Commit> commits = listFrom(tableClient, logPath, start)
+            .filter(fs -> FileNames.isCommitFile(getName(fs.getPath())))
+            .map(fs -> new Commit(FileNames.deltaVersion(fs.getPath()), fs.getModificationTime()));
+        return monotonizeCommitTimestamps(commits);
+    }
+
+    /**
+     * Makes sure that the commit timestamps are monotonically increasing with respect to commit
+     * versions. Requires the input commits to be sorted by the commit version.
+     */
+    private static List<Commit> monotonizeCommitTimestamps(CloseableIterator<Commit> commits) {
+        List<Commit> monotonizedCommits = new ArrayList<>();
+        long prevTimestamp = Long.MIN_VALUE;
+        long prevVersion = Long.MIN_VALUE;
+        while (commits.hasNext()) {
+            Commit newElem = commits.next();
+            assert(prevVersion < newElem.version); // Verify commits are ordered
+            if (prevTimestamp >= newElem.timestamp) {
+                logger.warn(
+                    "Found Delta commit {} with a timestamp {} which is greater than the next " +
+                        "commit timestamp {}.",
+                    prevVersion,
+                    prevTimestamp,
+                    newElem.timestamp);
+                newElem = new Commit(newElem.version, prevTimestamp + 1);
+            }
+            monotonizedCommits.add(newElem);
+            prevTimestamp = newElem.timestamp;
+            prevVersion = newElem.version;
+        }
+        return monotonizedCommits;
+    }
+
+    /** Returns the latest commit that happened at or before {@code timestamp} */
+    private static Optional<Commit> lastCommitBeforeTimestamp(
+            List<Commit> commits, long timestamp) {
+        int i = -1;
+        while (i + 1 < commits.size() && commits.get(i + 1).timestamp <= timestamp) {
+            i ++;
+        }
+        if (i < 0) { // none existed
+            return Optional.empty();
+        } else {
+            return Optional.of(commits.get(i));
+        }
+    }
+
+    private static class Commit {
```

Comment:
We should define an equals and a hashcode method. It would be error-prone to allow future develops to accidentally think that `Commit1(A,B)` != `Commit2(A,B)` just because the instances are different.

Reply:
Sure